### ISSUE TITLE
This type parsing error and incorrect type inference

### ISSUE THREAD

**Issue by stephanedr** [Microsoft/TypeScript#9744]
**TypeScript Version:**  2.0.0 (I suppose beta, though tsc -v just reports 2.0.0)

**Code**

``` ts
function test(e: HTMLElement): void {
    e.onclick = (this: HTMLElement, event) => { }; // error TS1005: ')' expected.
                                                   // error TS1005: ';' expected.
                                                   // error TS1128: Declaration or statement expected.

    e.onclick = function(this: HTMLElement, event) { }; // error TS7006: Parameter 'event' implicitly has an 'any' type
}
```

**Expected behavior:**
1st: no parsing error.
2nd: event: MouseEvent inference, as for `e.onclick = function(event) { };`

**Actual behavior:**

**Comment by RyanCavanaugh** [comment]
This is still going to be an error because it doesn't make sense for an arrow function to specify its `this` parameter (it gets the outer `this`), but we should give a semantic error instead of a parse error

**Comment by RyanCavanaugh** [comment]
Regarding the second thing, this is because we don't contextually type parameters when *any* parameter has a type annotation. There's a bug on that somewhere but I can't find it at the moment; we should revisit it (if it's closed) at the very least for the case where the only annotated parameter is the `this` the parameter

**Comment by mattmccutchen** [comment]
Let me add on to this issue (it doesn't seem worth filing another): a `this` parameter on a generic arrow function parses but has no effect:
```ts
let f = <T>(this: number) => {  // no error
    let that = this;  // resolved the same way as if `this: number` were not specified;
                      // may be an error depending on context and compiler options
};
```

**Comment by mhegazy** [comment]
thanks @mattmccutchen!
